csharp/src/Apache.Arrow/Arrays/BinaryArray.cs

```diff
@@ -66,87 +66,158 @@ protected BuilderBase(IArrowType dataType)
+            /// <returns>Returns the builder (for fluent-style composition).</returns>
             public TBuilder Append(ReadOnlySpan<byte> span)
             {
-                ValueOffsets.Append(Offset);
                 ValueBuffer.Append(span);
                 ValidityBuffer.Append(true);
                 Offset += span.Length;
+                ValueOffsets.Append(Offset);
                 return Instance;
             }
 
-            public TBuilder AppendRange(IEnumerable<byte[]> values)
+            /// <summary>
+            /// Append a value, consisting of an enumerable collection of bytes, to the array.
+            /// </summary>
+            /// <remarks>
+            /// Note that this method appends a single value, which may consist of arbitrarily many bytes.  If multiple
+            /// values are to be added, use the <see cref="AppendRange(IEnumerable{byte})"/> method instead.
+            /// </remarks>
+            /// <param name="value">Enumerable collection of bytes to add.</param>
+            /// <returns>Returns the builder (for fluent-style composition).</returns>
+            public TBuilder Append(IEnumerable<byte> value)
             {
-                foreach (byte[] arr in values)
+                if (value == null)
                 {
-                    if (arr == null)
-                    {
-                        AppendNull();
-                        continue;
-                    }
-                    int len = ValueBuffer.Length;
-                    ValueOffsets.Append(Offset);
-                    ValueBuffer.Append(arr);
-                    ValidityBuffer.Append(true);
-                    Offset += ValueBuffer.Length - len;
+                    return AppendNull();
                 }
 
+                // Note: by looking at the length of the value buffer before and after, we avoid having to iterate
+                // through the enumerable multiple times to get both length and contents.
+                int priorLength = ValueBuffer.Length;
+                ValueBuffer.AppendRange(value);
+                int valueLength = ValueBuffer.Length - priorLength;
+                Offset += valueLength;
+                ValidityBuffer.Append(true);
+                ValueOffsets.Append(Offset);
                 return Instance;
             }
 
+            /// <summary>
+            /// Append an enumerable collection of single-byte values to the array.
+            /// </summary>
+            /// <remarks>
+            /// Note that this method appends multiple values, each of which is a single byte.  If a single value is
+            /// to be added, use the <see cref="Append(IEnumerable{byte})"/> method instead.
+            /// </remarks>
+            /// <param name="values">Single-byte values to add.</param>
+            /// <returns>Returns the builder (for fluent-style composition).</returns>
             public TBuilder AppendRange(IEnumerable<byte> values)
```

Comment:
This `AppendRange(IEnumerable<byte>)` method previously did the "wrong thing": it was called `AppendRange()`, but it didn't add a range of values.  Instead, it added a single value.

This method now does what one would expect when calling `AppendRange()`, and adds a range of values (each of which is a single byte).  The original method behaviour is now available via `Append(IEnumerable<byte>)` instead.